### PR TITLE
#1780 add children registation feature

### LINKED ISSUE #1780: Register children into club
Design UI
Add new table(s) DB
Models
Develop Repository
Service
Controller
DTO + mapping

### PR DESCRIPTION
**Pull request by mikhalov**
The newly developed functionality includes several key services that significantly improve the user interaction process with the platform.

**Adding a child to the user:** The ``ChildService`` allows a user to create a child's profile linked to their account.

**Registration of a child in the club:** The new ``ClubRegistrationService`` allows children to be registered in clubs. With this service, the user can create an application for registration in the club. This process also includes validation and checking of active registration.

**Registration of a user-child in the club:** This feature allows a user-child to create their own profile and register in the club. It is designed for situations when a child independently uses the service and wants to participate in club activities.

**User registration in the club:** Similar to the registration of a child in the club, the user can also register in the club.

**Management of registrations:** The ``ClubRegistrationService`` also allows obtaining information about all unapproved registrations, canceling registrations, and approving registrations. This functionality is important for the role of a manager who can manage the registration process for the club.
**Additionally**, the user can view all their submitted applications.

### CODE REVIEW

File: src/main/java/com/softserve/teachua/controller/ClubController.java [hunk 1]
```diff
@@ -106,8 +107,10 @@ public List<ClubResponse> getClubsByCenterId(@PathVariable Long id) {
      */
     @AllowedRoles({RoleData.ADMIN, RoleData.MANAGER})
     @PostMapping("/club")
-    public SuccessCreatedClub addClub(@Valid @RequestBody ClubProfile clubProfile) {
-        return clubService.addClub(clubProfile);
+    public ResponseEntity<SuccessCreatedClub> addClub(@Valid @RequestBody ClubProfile clubProfile) {
+        SuccessCreatedClub club = clubService.addClub(clubProfile);
+
+        return ResponseEntity.ok(club);
```
Comment (vitalik12345678) [inline src/main/java/com/softserve/teachua/controller/ClubController.java]: Response status should be 201(CREATED)

File: src/main/java/com/softserve/teachua/controller/ClubRegistrationController.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        return ResponseEntity.ok(applications);
+    }
+
+    @AllowedRoles(RoleData.USER)
+    @PostMapping("/club-registration")
+    public ResponseEntity<List<ClubRegistrationResponse>> addClubRegistration(
+            @Valid @RequestBody ClubRegistrationRequest clubRegistrationRequest) {
+        List<ClubRegistrationResponse> response = clubRegistrationService.create(clubRegistrationRequest);
+
+        return ResponseEntity.ok(response);
```
Comment (vitalik12345678) [inline src/main/java/com/softserve/teachua/controller/ClubRegistrationController.java]: 201 or CREATED

File: src/main/java/com/softserve/teachua/controller/ClubRegistrationController.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @AllowedRoles(RoleData.USER)
+    @PostMapping("/club-registration/user")
+    public ResponseEntity<UserClubRegistrationResponse> addClubRegistrationForUser(
+            @Valid @RequestBody UserClubRegistrationRequest userClubRegistrationRequest) {
+        UserClubRegistrationResponse response = clubRegistrationService.create(userClubRegistrationRequest);
+
+        return ResponseEntity.ok(response);
```
Comment (vitalik12345678) [inline src/main/java/com/softserve/teachua/controller/ClubRegistrationController.java]: The same

File: src/main/java/com/softserve/teachua/service/impl/ChildServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ChildServiceImpl implements ChildService {
+    private final ChildRepository childRepository;
+    private final DtoConverter dtoConverter;
+    private final UserService userService;
+
+    @Override
```
Comment (vitalik12345678) [inline src/main/java/com/softserve/teachua/service/impl/ChildServiceImpl.java:27]: @Transactional

File: src/main/java/com/softserve/teachua/service/impl/ChildServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        child.setParent(user);
+        log.debug("Set parent {}", child);
+
+        child = childRepository.save(child);
+
+        return dtoConverter.convertToDto(child, ChildResponse.class);
+    }
+
+    @Override
+    public Child getById(Long id) {
```
Comment (vitalik12345678) [inline src/main/java/com/softserve/teachua/service/impl/ChildServiceImpl.java:44]: @TRansactional(readOnly = true ) or on class level

File: src/main/java/com/softserve/teachua/service/impl/ClubRegistrationServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+@Service
+@RequiredArgsConstructor
+public class ClubRegistrationServiceImpl implements ClubRegistrationService {
+    private final ClubRegistrationRepository clubRegistrationRepository;
+    private final ChildService childService;
+    private final UserService userService;
+    private final ClubService clubService;
+    private final DtoConverter dtoConverter;
+
+    @Override
```
Comment (vitalik12345678) [inline src/main/java/com/softserve/teachua/service/impl/ClubRegistrationServiceImpl.java:42]: @Transactional problems

File: src/main/java/com/softserve/teachua/service/impl/ClubRegistrationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.function.Function;
+import com.softserve.teachua.service.UserService;
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ClubRegistrationServiceImpl implements ClubRegistrationService {
```
Comment (vitalik12345678) [inline src/main/java/com/softserve/teachua/service/impl/ClubRegistrationServiceImpl.java:35]: Need add loging and Transacntional

File: src/main/java/com/softserve/teachua/service/impl/ClubRegistrationServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        var dto = new UnapprovedClubRegistration();
+
+        return unapprovedClubRegistrations.stream()
+                .map(ucr -> dtoConverter.convertToDto(ucr, dto))
+                .toList();
+    }
+
+    @Override
+    public List<FullClubRegistration> getAllByManagerId(Long managerId) {
+        var clubRegistrations = clubRegistrationRepository
```
Comment (vitalik12345678) [inline src/main/java/com/softserve/teachua/service/impl/ClubRegistrationServiceImpl.java]: user var or Class name , not together

**Review by softservedata — APPROVED** [review]

**Pull request merged**